delta/plugins/storage/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/storage/storages/operations/s3/client/S3StorageClient.scala

```diff
@@ -72,7 +88,75 @@ object S3StorageClient {
     override def headObject(bucket: String, key: String): IO[HeadObjectResponse] =
       client.headObject(HeadObjectRequest.builder().bucket(bucket).key(key).checksumMode(ChecksumMode.ENABLED).build)
 
-    override def underlyingClient: S3AsyncClientOp[IO] = client
+    override def objectExists(bucket: String, key: String): IO[Boolean] = {
+      headObject(bucket, key)
+        .redeemWith(
+          {
+            case _: NoSuchKeyException => IO.pure(false)
+            case e                     => IO.raiseError(e)
+          },
+          _ => IO.pure(true)
+        )
+    }
+
+    override def uploadFile(
+        fileData: Stream[IO, Byte],
+        bucket: String,
+        key: String,
+        algorithm: DigestAlgorithm
+    ): IO[UploadMetadata] = {
+      for {
+        fileSizeAcc <- Ref.of[IO, Long](0L)
+        digest      <- fileData
+                         .evalTap(_ => fileSizeAcc.update(_ + 1))
```

Comment:
The size could be retrieved by querying S3.
If it is not returned by the put operation, then we can follow the put by a headObject like the registering operation